### PR TITLE
Only run task once for #53

### PR DESCRIPTION
**Pull request by RossHammer**
This should implement the feature from #53

Everything should be done and working but I'm not sure what to do with the taskfile version. Should I add the code+docs to increment the version to 3.1? I think I'm also missing a couple checks for setting the run property on versions before this.

### CODE REVIEW

**Comment by andreynering** [comment]
Hi @RossHammer!

I haven't reviewed the code yet, but in a quick look the approach seems interesting!

> any feedback on how I should handle incrementing the version of the taskfile if at all? I would be nice too get this added

The this change is backward compatible (i.e. the default option keeps the current behavior) you don't need to do anything

**Comment by RossHammer** [comment]
Awesome thanks! I have updated it to handle v2 properly and it should be done then.

**Comment by dolfelt** [comment]
Is there any movement on this? I'm trying to move to `go-task` from our own custom Go task runner that handles this use case. Nice work @RossHammer!

File: internal/hash/hash.go [hunk 1]
```diff
@@ -0,0 +1,23 @@
+package hash
+
+import (
+	"fmt"
+
+	"github.com/go-task/task/v3/taskfile"
+	"github.com/mitchellh/hashstructure"
```
Comment (andreynering) [inline internal/hash/hash.go:7]: Import grouping should be ordered by: stdlib, libs, Task packages.

```suggestion
	"github.com/mitchellh/hashstructure"

	"github.com/go-task/task/v3/taskfile"
```

**Review by andreynering — COMMENTED** [review]
Hi @RossHammer,

First of all, sorry for the long wait. I liked the approach and this is feature is important to many people, so let's proceed!

I'd like to know from you if you're still interested in finishing this PR; otherwise I'll take over fixing the conflicts + making a few changes before shipping.

Either way, I added a few comments about minor stuff we need to change.

**Comment by RossHammer** [comment]
I can take a look at the PR but I won't be able to do it for a a few weeks. Please feel free to take it over if you have a chance to look before I get to it but I'll see what I can do!

File: task.go [hunk 2]
```diff
@@ -205,6 +207,13 @@ func (e *Executor) Setup() error {
 		}
 	}
 
+	if v < 3 && e.Taskfile.Run != "" {
```
[the remark this answers is not in the document]
Comment (RossHammer) [reply task.go]: I have updated everything to start using 3.7 instead

File: task.go [hunk 3]
```diff
@@ -399,3 +426,25 @@ func getEnviron(t *taskfile.Task) []string {
 	}
 	return environ
 }
+
+func (e *Executor) startExecution(innerCtx context.Context, t *taskfile.Task) (bool, context.Context, context.CancelFunc, error) {
```
[the remark this answers is not in the document]
Comment (RossHammer) [reply task.go]: I did change it a bit, parts are better but I'm sure there is still room for improvement. Hopefully it is at least more clear!

**Comment by RossHammer** [comment]
I'm not exactly sure what this is going to do with the watch flag, I have never really used it before. I would suspect this will just never run the `once` and `when_changed` types again. Is this what we want? If not I can try and look into changing it!

**Comment by andreynering** [comment]
Hi @RossHammer,

For some reason GitHub didn't automatically marked this PR as merged, but I just manually merged this into master with a small post commit: 8aa983257da6da096a621200940df8f3d39be065.

This worked awesomely well in my tests. Thanks a lot for your hard work and patience! ❤️

**Comment by Porges** [comment]
When I have some `vars` at top level and I change them the task doesn’t re-run. Can I confirm that it is expected that this doesn’t  account for vars that are not _explicitly_ passed to the task?

**Comment by andreynering** [comment]
Hi @Porges,

An example would be welcome for me to understand better the question.

But in short, with `when_changed`, a hash of the task is calculated to know if it should run again. This means that only vars that are concatenated in any prop of the task are taken in consideration here.

**Comment by Porges** [comment]
@andreynering I think my problem is actually not really related to `when_changed` but more that var hashes aren't taken into account when determining if the task should re-run.

e.g.
```yaml
version: '3.7'

run: when_changed

vars:
  VERSION: 1

tasks:
  generate-file:
    sources:
    - base.txt
    cmds:
    - <base.txt sed -e 's/VERSION/{{.VERSION}}/g' > versioned.txt
```

This only calculates the `base.txt` hash and doesn’t re-run when `VERSION` is changed.

**Pull request closed without merging**